wrap_function: Allow general function wrapping to facilitate consistent pybind workflows.

EricCousineau-TRI opened the pull request:
This implements functionality to enable a potential solution for #7793, and to provide workarounds for https://github.com/pybind/pybind11/issues/1241 (callbacks and virtual function overrides - #7838).

The prototype for `const`-proxying, connected with `pybind`is here:
[cpp_const_pybind.h](https://github.com/EricCousineau-TRI/repro/blob/242807deabbb71801a8bf73f193292c7adeb2e42/python/bindings/pymodule/const/cpp_const_pybind.h)
[cpp_const_pybind_test.py](https://github.com/EricCousineau-TRI/repro/blob/242807deabbb71801a8bf73f193292c7adeb2e42/python/bindings/pymodule/const/test/cpp_const_pybind_test.py)

This is a tad hefty with both functionality and the unittest, but I feel the unittest explains usage most easily.
If desired, I can remove the nested callback tests and save that for a follow-up PR.

NOTE: I found `clang-tidy` to be a little too picky.
I want to ensure that this is tested for mutable rvalue references, and there was no way to disable this for the entire file, from within the file.
Additionally, its desired for `std::function<>` signatures is weird - it doesn't want spaces?

Because of these, and because I feel like `NOLINTNEXTLINE(...) <explanation>` is verbose and clutters the code, I've used `NOLINT`.

```
$ drake_cloc.py 
Computing difference between $(git merge-base master HEAD) (f3a7a38) and 'HEAD' (f3f8915).

Category            added  modified  removed  
----------------------------------------------
code                379    0         0        
comments            123    0         0        
blank               91     0         0        
----------------------------------------------
TOTAL               593    0         0    
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7841)
<!-- Reviewable:end -->

EricCousineau-TRI commented:
+@jwnimmer-tri for feature review, please.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-:-L3_1bURNQWQr36DHX1v:b-qf68jg)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 3 files at r1.
Review status: 2 of 3 files reviewed at latest revision, 20 unresolved discussions.

---

*[bindings/pydrake/util/wrap_function.h, line 9 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_cS-plVSq6nwJdUL5:-L3_cS-plVSq6nwJdUL6:bf21tnm) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L9)):*
> ```objc
> namespace drake {
> 
> namespace detail {
> ```

Probably the `drake::detail` namespace is getting a little crowded.  Maybe consider `wrap_function_detail` since only ~4 places have to spell it out?

---

*[bindings/pydrake/util/wrap_function.h, line 20 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_buznIF9LGAA5J7NK:-L3_buznIF9LGAA5J7NL:bmlmcob) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L20)):*
> ```objc
> };
> 
> // Implementation for inferring `function_info<>` from a generic `Func` object.
> ```

BTW This didn't help me much.  Is this basically just the factory method for `struct function_info`, without anything else fancy going on?  If so, just say that?  Or rename it `make_function_info` or something?

---

*[bindings/pydrake/util/wrap_function.h, line 95 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_UY81dLaXoCpCo8JP:-L3_UY81dLaXoCpCo8JQ:btjqj6j) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L95)):*
I don't understand what this comment means.

(...later...)

Oh I see; much later on we specialize this.  It would have helped me if comment here said something more like this:

> By default `wrap_arg` is the same as `wrap_arg_tpl`.  However, below we specialize it for the case when `T = std::function`.

---

*[bindings/pydrake/util/wrap_function.h, line 114 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_XAA519m7aKfCa0o1:-L3_XAA6JK8ffJNpuCJX:b-koul65) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L114)):*
> ```objc
>   // Convenience wrapper.
>   template <typename T>
>   using wrap_arg_t = typename wrap_arg_t_impl<T, void>::type;
> ```

The number of things within this struct named starting with `wrap_arg` is too difficult to follow (and a little bit inconsistent).  Since most flavors of these names are just a single-use intermediate step along the way to `wrap_arg_t`, it seems like we could afford to use more descriptive names for them (maybe not even not needing the `wrap_arg` prefix on some of them?

---

*[bindings/pydrake/util/wrap_function.h, line 128 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3__t29rqd8JtvJeX3g:-L3__t29rqd8JtvJeX3h:b7ga8yb) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L128)):*
> ```objc
>       std::enable_if_t<enable_wrap_output<Return>, void*> = {}) {
>     // N.B. Since we do not use the `mutable` keyword with this lambda,
>     // any functors passed in *must* have provide `operator()(...) const`.
> ```

BTW typo "have provide"

---

*[bindings/pydrake/util/wrap_function.h, line 153 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_Xny1Oudl48Obns3-:-L3_Xny2H-4p4f3sjhCD:b327y4h) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L153)):*
> ```objc
>   }
> 
>   // Specialization for callbacks of the form `std::function<>`.
> ```

FYI I am not 100% sure but I _probably_ would have had an easier time reading this if all of the wrap_arg specializations happened before `run`.  Having `run` buried in the middle was a little bit difficult.

---

*[bindings/pydrake/util/wrap_function.h, line 157 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_a9A_wJ5mCx2Q8At0:-L3_a9AaGIGzAL9oWnU0:bdndd82) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L157)):*
> ```objc
>   // that complicates the details for a relatively low ROI.
>   template <typename Return, typename ... Args>
>   struct wrap_arg<const std::function<Return (Args...)>&> {  // NOLINT
> ```

Is the NOLINT just for the space in `Return (Args` signature?  The http://en.cppreference.com/w/cpp/utility/functional/function spells it without the space, and omitting the space seems fine by me, maybe even better?  I don't mind the bare NOLINT's in the test, but in a prod code header it would be nice if we could avoid it.

(Ditto for the earlier NOLINT's in this file, which I think were for the same reason.)

---

*[bindings/pydrake/util/wrap_function.h, line 210 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_SgWfmU-cpXm1dkMe:-L3_SgWfmU-cpXm1dkMf:b-qhtesg) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L210)):*
> ```objc
>   // N.B. `T` rather than `T&&` is used as arguments here as it behaves well
>   // with primitve types, such as `int`.
> };
> ```

FYI Consider moving this example struct to be _after_ the next one that would use it; I tried to start reading the PR here (at the first non-detail thing) and was confused what we were trying to do.   Of if the order says put, maybe the class overview here could immediately say "Example helper for `WrapFunction` below." which would have clued me in to look there.

---

*[bindings/pydrake/util/wrap_function.h, line 213 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_awgf-fOmW2Zjgq3D:-L3_awgf-fOmW2Zjgq3E:b-f6cq6t) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L213)):*
> ```objc
> 
> /// Wraps a functions arguments and its return value (if non-void) based on
> /// `wrap_arg_tpl`. This will also recursively handle wrapping /unwrapping
> ```

BTW Saying that wrapping is about the _arguments_ would help here -- as in "decorating and undecorating the function arguments" or something.

---

*[bindings/pydrake/util/wrap_function.h, line 219 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_ZWaEyXfIkloZnpJs:-L3_ZWaFT39hjMeMSONQ:bc2ehyv) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L219)):*
> ```objc
> ///   `wrap_arg_tpl<T>::wrap(auto arg, ...)` and
> ///   `wrap_arg_tpl<T>::unwrap(auto arg_wrapped, ...)`. See `wrap_arg_default`
> ///   for a simple implementation.
> ```

Either here (or possibly in wrap_arg_default) we need to clarify the directionality of "wrap" and "unwrap".  The wrap functor takes in values in the same domain or range as Func, and produces values in the domain or range of WrapFunction(Func).

---

*[bindings/pydrake/util/wrap_function.h, line 231 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_U0uHfmTJT-_IsXvu:-L3_U0uHfmTJT-_IsXvv:b-eoo5lo) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L231)):*
> ```objc
> ///   N.B. Construct a `std::function<>` from this if you encounter inference
> ///   issues downstream of this method.
> template <template <typename...> class wrap_arg_tpl, typename Func>
> ```

The name `wrap_arg_tpl` is a bit obscure on its own.  Is it really something like a "policy", perhaps?  (Maybe `wrap_arg_policy` here?)

This name also has the `tpl` abbreviation nit.  Also, I don't understand why some names are `_tpl` and some names are `_t ` -- both suffixes seem to indicate that its a template template parameter?

---

*[bindings/pydrake/util/wrap_function.h, line 239 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_T7WzZpCbOOqpblmg:-L3_T7WzZpCbOOqpblmh:b-ulnko0) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L239)):*
> ```objc
> }
> 
> }  // namespace drake
> ```

Do we want these utilities directly in `drake` when they're not in `//common`?  Would putting this in `pydrake` here be more consistent with the surrounding code in `util`?

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 9 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_SPauewc5UlBjNPzm:-L3_SPauewc5UlBjNPzn:bbyr63d) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/test/wrap_function_test.cc#L9)):*
> ```cpp
> 
> // @note Some of these statements violate style guide. That is intended, as
> // this should (a) test references for use in `pybind`, and (b) clang-tidy does
> ```

BTW Here I think you mean `cpplint` (not `clang-tidy`), which is the thing we have integrated into bazel that complains about formatting.

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 23 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_dwX5BSPriV7O7dYW:-L3_dwX65vSpfRESAfGk:b-oi7ahy) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/test/wrap_function_test.cc#L23)):*
> ```cpp
> }
> 
> #define WTEST(name) GTEST_TEST(WrapFunction, name)
> ```

This doesn't seem to serve any strong enough purpose to admit an exception to cppguide preprocessor rules?  Or at least, document why if so.

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 26 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_ePBnaajGGfJlwLW8:-L3_ePBnaajGGfJlwLW9:b-kdrdvs) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/test/wrap_function_test.cc#L26)):*
> ```cpp
> 
> // Function with `void` return type, `int` by value.
> void FunctionPtr(int value) {}
> ```

`Ptr` abbreviation (this also isn't technically isn't decayed to a pointer yet?)  (Maybe `void IntToVoid(int) {}`?)

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_eeEXp7oLS5zIzuvd:-L3_eeEXp7oLS5zIzuve:bqd0bqi) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/test/wrap_function_test.cc#L30)):*
> ```cpp
> WTEST(FunctionPtr) {
>   int value = 0;
>   WrapIdentity(FunctionPtr)(value);
> ```

BTW Should passing either `FunctionPtr` or `&FunctionPtr` both work?  Should both spellings be exercised in the test here?

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 31 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_etDXQHrBAv2QaAX6:-L3_etDXQHrBAv2QaAX7:b-7uover) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/test/wrap_function_test.cc#L31)):*
> ```cpp
>   int value = 0;
>   WrapIdentity(FunctionPtr)(value);
>   EXPECT_EQ(value, 0);
> ```

I don't understand the purpose of this assertion?

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 39 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_epMGL4pR1esN0NUJ:-L3_epMHyBxzb_IdIUS6:b-7uover) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/test/wrap_function_test.cc#L39)):*
> ```cpp
>   auto func_1_lambda = [](int value) {};
>   WrapIdentity(func_1_lambda)(value);
>   EXPECT_EQ(value, 0);
> ```

I don't understand the purpose of this assertion?

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 43 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_euCs2mjNtdgbj3-M:-L3_euCs2mjNtdgbj3-N:b-7uover) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/test/wrap_function_test.cc#L43)):*
> ```cpp
>   std::function<void (int)> func_1_func = func_1_lambda;  // NOLINT
>   WrapIdentity(func_1_func)(value);
>   EXPECT_EQ(value, 0);
> ```

I don't understand the purpose of this assertion?

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 143 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_gbC-VKguJWnIveZl:-L3_gbC-VKguJWnIveZm:b8k3n6j) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/test/wrap_function_test.cc#L143)):*
> ```cpp
> 
> 
> // Test a slightly complicated conversion mechanism.
> ```

Checkpoint.  I have to pause now, but I'll resume reading below tomorrow.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7841)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 1 of 3 files reviewed at latest revision, 20 unresolved discussions.

---

*[bindings/pydrake/util/wrap_function.h, line 9 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_cS-plVSq6nwJdUL5:-L3g0kpZqBcNN7J9Q6L8:b-ds1mzw) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L9)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Probably the `drake::detail` namespace is getting a little crowded.  Maybe consider `wrap_function_detail` since only ~4 places have to spell it out?
</blockquote></details>

Done. Placed in `drake::pydrake`.

---

*[bindings/pydrake/util/wrap_function.h, line 20 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_buznIF9LGAA5J7NK:-L3g0o0mEu66CcTvMFJm:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L20)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW This didn't help me much.  Is this basically just the factory method for `struct function_info`, without anything else fancy going on?  If so, just say that?  Or rename it `make_function_info` or something?
</blockquote></details>

Done.

---

*[bindings/pydrake/util/wrap_function.h, line 95 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_UY81dLaXoCpCo8JP:-L3g0rAUH2je6IZ52v-V:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L95)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I don't understand what this comment means.

(...later...)

Oh I see; much later on we specialize this.  It would have helped me if comment here said something more like this:

> By default `wrap_arg` is the same as `wrap_arg_tpl`.  However, below we specialize it for the case when `T = std::function`.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/wrap_function.h, line 114 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_XAA519m7aKfCa0o1:-L3g0sBHL6TAlzXWucSt:bxx1bp4) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L114)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The number of things within this struct named starting with `wrap_arg` is too difficult to follow (and a little bit inconsistent).  Since most flavors of these names are just a single-use intermediate step along the way to `wrap_arg_t`, it seems like we could afford to use more descriptive names for them (maybe not even not needing the `wrap_arg` prefix on some of them?
</blockquote></details>

Done.
The intent of the `_t` suffix is to mirror the C++14 helper types, e.g. [`decay_t<>`](http://en.cppreference.com/w/cpp/types/decay#Helper_types).

---

*[bindings/pydrake/util/wrap_function.h, line 128 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3__t29rqd8JtvJeX3g:-L3g1CF0PUtGfABLCzL_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L128)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW typo "have provide"
</blockquote></details>

Done.

---

*[bindings/pydrake/util/wrap_function.h, line 153 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_Xny1Oudl48Obns3-:-L3g1D1gZryU13yAZX8E:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L153)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI I am not 100% sure but I _probably_ would have had an easier time reading this if all of the wrap_arg specializations happened before `run`.  Having `run` buried in the middle was a little bit difficult.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/wrap_function.h, line 157 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_a9A_wJ5mCx2Q8At0:-L3g1Dt8K3JsWeXC7xLg:b-irhwzw) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L157)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Is the NOLINT just for the space in `Return (Args` signature?  The http://en.cppreference.com/w/cpp/utility/functional/function spells it without the space, and omitting the space seems fine by me, maybe even better?  I don't mind the bare NOLINT's in the test, but in a prod code header it would be nice if we could avoid it.

(Ditto for the earlier NOLINT's in this file, which I think were for the same reason.)
</blockquote></details>

Done.
I like it less without the space, but will stick with what cppreference uses, then. Thanks for pointing that out!

---

*[bindings/pydrake/util/wrap_function.h, line 210 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_SgWfmU-cpXm1dkMe:-L3g2AnRIVVFoKPF8QWt:bqrvjf9) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L210)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Consider moving this example struct to be _after_ the next one that would use it; I tried to start reading the PR here (at the first non-detail thing) and was confused what we were trying to do.   Of if the order says put, maybe the class overview here could immediately say "Example helper for `WrapFunction` below." which would have clued me in to look there.
</blockquote></details>

Moved - will add reference in follow-up.

---

*[bindings/pydrake/util/wrap_function.h, line 219 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_ZWaEyXfIkloZnpJs:-L3g2GC9I999mNqbEOUr:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L219)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Either here (or possibly in wrap_arg_default) we need to clarify the directionality of "wrap" and "unwrap".  The wrap functor takes in values in the same domain or range as Func, and produces values in the domain or range of WrapFunction(Func).
</blockquote></details>

Done.

---

*[bindings/pydrake/util/wrap_function.h, line 231 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_U0uHfmTJT-_IsXvu:-L3g2RqsxVbTnPbzIb-d:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L231)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The name `wrap_arg_tpl` is a bit obscure on its own.  Is it really something like a "policy", perhaps?  (Maybe `wrap_arg_policy` here?)

This name also has the `tpl` abbreviation nit.  Also, I don't understand why some names are `_tpl` and some names are `_t ` -- both suffixes seem to indicate that its a template template parameter?
</blockquote></details>

Done.

---

*[bindings/pydrake/util/wrap_function.h, line 239 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_T7WzZpCbOOqpblmg:-L3g2Som99ENB9TRogSB:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L239)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Do we want these utilities directly in `drake` when they're not in `//common`?  Would putting this in `pydrake` here be more consistent with the surrounding code in `util`?
</blockquote></details>

Done.

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 9 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_SPauewc5UlBjNPzm:-L3g2TgEMsA_3BL64j2O:b-y26q6t) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/test/wrap_function_test.cc#L9)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Here I think you mean `cpplint` (not `clang-tidy`), which is the thing we have integrated into bazel that complains about formatting.
</blockquote></details>

Done. Removed, given that cppreference uses it without spaces.

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 23 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_dwX5BSPriV7O7dYW:-L3g2Y4hdP0XhbvK1AIc:b-hit5cg) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/test/wrap_function_test.cc#L23)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This doesn't seem to serve any strong enough purpose to admit an exception to cppguide preprocessor rules?  Or at least, document why if so.
</blockquote></details>

Done. Sorry, had forgotten to remove that!

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 26 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_ePBnaajGGfJlwLW8:-L3g2b4wft64WG0n2_zB:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/test/wrap_function_test.cc#L26)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

`Ptr` abbreviation (this also isn't technically isn't decayed to a pointer yet?)  (Maybe `void IntToVoid(int) {}`?)
</blockquote></details>

Done.

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_eeEXp7oLS5zIzuvd:-L3g2bkkxmrvC4jpGSJe:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/test/wrap_function_test.cc#L30)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Should passing either `FunctionPtr` or `&FunctionPtr` both work?  Should both spellings be exercised in the test here?
</blockquote></details>

Done.

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 31 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_etDXQHrBAv2QaAX6:-L3g2cspqTPwzTQp3aSG:b-wnfmk9) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/test/wrap_function_test.cc#L31)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I don't understand the purpose of this assertion?
</blockquote></details>

Done. Removed.

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 39 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_epMGL4pR1esN0NUJ:-L3g2doycFFyc1i9Mdk6:b-wnfmk9) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/test/wrap_function_test.cc#L39)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I don't understand the purpose of this assertion?
</blockquote></details>

Done. Removed.

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 43 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_euCs2mjNtdgbj3-M:-L3g2enlCanJ953CoAce:b-wnfmk9) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/test/wrap_function_test.cc#L43)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I don't understand the purpose of this assertion?
</blockquote></details>

Done. Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7841)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
+@soonho-tri for platform review, please.  (I still have a little more feature review to do, but I think this is mostly ready.)

---

Reviewed 1 of 2 files at r2.
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[bindings/pydrake/util/wrap_function.h, line 114 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_XAA519m7aKfCa0o1:-L3hJcfy_GvyWqvnK7IL:b-y8nt62) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L114)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Done.
The intent of the `_t` suffix is to mirror the C++14 helper types, e.g. [`decay_t<>`](http://en.cppreference.com/w/cpp/types/decay#Helper_types).
</blockquote></details>

OK That makes sense; somehow I didn't connect the dots to infer that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-:-L3hUfHM1_ndCO3y25jh:beo61it)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 157 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3hVJp5ZvdwzOEWU6Tk:-L3hVJp5ZvdwzOEWU6Tl:b-fkgwlk) ([raw file](https://github.com/robotlocomotion/drake/blob/cd166cfb24dc56df72bd71a648fddfe5179545ff/bindings/pydrake/util/test/wrap_function_test.cc#L157)):*
> ```cpp
> template <typename T>
> struct ptr {
>   // Ensure that this is not being used in lieu of `ptr` (ensure that
> ```

BTW Do you mean "lieu of `const_ptr`"?

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 231 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3hZqsp1eLLy21TOo2s:-L3hZqsp1eLLy21TOo2t:b2v4tpj) ([raw file](https://github.com/robotlocomotion/drake/blob/cd166cfb24dc56df72bd71a648fddfe5179545ff/bindings/pydrake/util/test/wrap_function_test.cc#L231)):*
> ```cpp
> }
> 
> template <typename Expected, typename Actual>
> ```

FYI With `EXPECT_EQ` we typically (and this code also does) the order (actual, expected).

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 270 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3h__mq3cESeUdQZ34N:-L3h__mq3cESeUdQZ34O:b2r2kk) ([raw file](https://github.com/robotlocomotion/drake/blob/cd166cfb24dc56df72bd71a648fddfe5179545ff/bindings/pydrake/util/test/wrap_function_test.cc#L270)):*
> ```cpp
>     (void)Dummy{(check_type<Args, IArgs>(), 0)...};
>   }
> };
> ```

FYI This could move lower (by one test case) -- its not used by the ChangeBasic test.  (Or one higher, to be next to its helper `check_type`.)

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 332 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3hb8XnRvZPRN8dRc6h:-L3hb8XnRvZPRN8dRc6i:b-fiy8rx) ([raw file](https://github.com/robotlocomotion/drake/blob/cd166cfb24dc56df72bd71a648fddfe5179545ff/bindings/pydrake/util/test/wrap_function_test.cc#L332)):*
> ```cpp
> Callback ChangeCallback(const Callback&) { throw std::runtime_error("Dummy"); }
> 
> GTEST_TEST(WrapFunction, ChangeCallback) {
> ```

When I read the API documentation for `WrapFunction`, I did not get the message that when function arguments are functions, they are recursively wrapped as well.  That seems like it should be a part of the API docs and contract.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-:-L3hbhkNbkEFzFTkZoS5:bnfp4nl)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 1 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[bindings/pydrake/util/wrap_function.h, line 114 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_XAA519m7aKfCa0o1:-L3j3b2oHzYNnSwg1dKv:bsz5spv) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L114)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

OK That makes sense; somehow I didn't connect the dots to infer that.
</blockquote></details>

OK Yeah, I had screwed up that convention in that I used `wrap_arg_t = wrap_arg_type_impl::type`, whereas the `std` helper types are usually of the form like `decay_t = decay::type`, probably why it was hard to connect.

---

*[bindings/pydrake/util/wrap_function.h, line 213 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3_awgf-fOmW2Zjgq3D:-L3j6_RNMtU-p9DiGR2r:bj40bdf) ([raw file](https://github.com/robotlocomotion/drake/blob/f3f89158ec89a49756217f7821e4c70c460d4b3f/bindings/pydrake/util/wrap_function.h#L213)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Saying that wrapping is about the _arguments_ would help here -- as in "decorating and undecorating the function arguments" or something.
</blockquote></details>

Done. A little tentative about "decorating", because that implies that some part of the type may stay behind. In theory, a user could completely redirect types.

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 157 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3hVJp5ZvdwzOEWU6Tk:-L3j7AWf1zpuVRSeMRaW:b-qsx7uj) ([raw file](https://github.com/robotlocomotion/drake/blob/cd166cfb24dc56df72bd71a648fddfe5179545ff/bindings/pydrake/util/test/wrap_function_test.cc#L157)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Do you mean "lieu of `const_ptr`"?
</blockquote></details>

Done. Oops!

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 231 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3hZqsp1eLLy21TOo2s:-L3j7BeOtX5L8wlVfQQN:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/cd166cfb24dc56df72bd71a648fddfe5179545ff/bindings/pydrake/util/test/wrap_function_test.cc#L231)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI With `EXPECT_EQ` we typically (and this code also does) the order (actual, expected).
</blockquote></details>

Done.

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 270 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3h__mq3cESeUdQZ34N:-L3j7GfL42dmd30P27G4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/cd166cfb24dc56df72bd71a648fddfe5179545ff/bindings/pydrake/util/test/wrap_function_test.cc#L270)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI This could move lower (by one test case) -- its not used by the ChangeBasic test.  (Or one higher, to be next to its helper `check_type`.)
</blockquote></details>

Done.

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 332 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3hb8XnRvZPRN8dRc6h:-L3j7IC_nd03jEcvCxwY:brh686x) ([raw file](https://github.com/robotlocomotion/drake/blob/cd166cfb24dc56df72bd71a648fddfe5179545ff/bindings/pydrake/util/test/wrap_function_test.cc#L332)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

When I read the API documentation for `WrapFunction`, I did not get the message that when function arguments are functions, they are recursively wrapped as well.  That seems like it should be a part of the API docs and contract.
</blockquote></details>

Done. Updated the API docs / contract.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7841)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7841)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Checkpoint (still processing the tests). This template soup is pretty dense.

---

Reviewed 1 of 3 files at r1, 1 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[bindings/pydrake/util/wrap_function.h, line 31 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3jLAYBAft1e6kvHwWE:-L3jLAYCk58xBqnf5N36:bfvn7l3) ([raw file](https://github.com/robotlocomotion/drake/blob/d3e9d8a621d47135fd5cc3c4dc637fa176820cb3/bindings/pydrake/util/wrap_function.h#L31)):*
> ```objc
> // lambda objects. It does *not* distinguish among other types.
> template <typename Func, typename T = void>
> using enable_if_lambda_t =
> ```

BTW, isn't it the same as the following?
```c++
template <typename Func, typename T = void>
using enable_if_lambda_t =
    std::enable_if_t<!std::is_function<std::decay_t<Func>>::value, T>;
```

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 43 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3j26oaph_ef5l9D2Vw:-L3j26oaph_ef5l9D2Vx:bnnsach) ([raw file](https://github.com/robotlocomotion/drake/blob/cd166cfb24dc56df72bd71a648fddfe5179545ff/bindings/pydrake/util/test/wrap_function_test.cc#L43)):*
> ```cpp
>   int value{0};
>   auto func_1_lambda = [](int value) {};
>   WrapIdentity(func_1_lambda)(value);
> ```

BTW, no need to have `EXPECT_EQ` for this one and the following one?

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 147 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3jUGjQzqwoOQnP8aMz:-L3jUGjQzqwoOQnP8aN-:bh1wx4f) ([raw file](https://github.com/robotlocomotion/drake/blob/d3e9d8a621d47135fd5cc3c4dc637fa176820cb3/bindings/pydrake/util/test/wrap_function_test.cc#L147)):*
> ```cpp
> // Test a slightly complicated conversion mechanism.
> 
> // Constructible only via brackets (no implicit conversions).
> ```

Not sure how this comment is related to the following code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-:-L3jW1KI-wqiS3KN1QVV:bboqbzz)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 251 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3nC8j_GSUUVOnGS7Wn:-L3nC8j_GSUUVOnGS7Wo:b-hps881) ([raw file](https://github.com/robotlocomotion/drake/blob/d3e9d8a621d47135fd5cc3c4dc637fa176820cb3/bindings/pydrake/util/test/wrap_function_test.cc#L251)):*
> ```cpp
>     check_type<ReturnActual, ReturnExpected>();
>     using Dummy = int[];
>     (void)Dummy{(check_type<ArgsActual, ArgsExpected>(), 0)...};
> ```

:-) I've spent some time to figure out how it works (comma operator).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-:-L3nBH25akK6Ays_Pbsk:bnfp4nl)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 1 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[bindings/pydrake/util/wrap_function.h, line 31 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3jLAYBAft1e6kvHwWE:-L3ne3pLuar_b09WbMG0:b-gu0rze) ([raw file](https://github.com/robotlocomotion/drake/blob/d3e9d8a621d47135fd5cc3c4dc637fa176820cb3/bindings/pydrake/util/wrap_function.h#L31)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, isn't it the same as the following?
```c++
template <typename Func, typename T = void>
using enable_if_lambda_t =
    std::enable_if_t<!std::is_function<std::decay_t<Func>>::value, T>;
```
</blockquote></details>

Done. Oops! Had originally wrapped the value, but then switched to the enable if.

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 43 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3j26oaph_ef5l9D2Vw:-L3nd7P6ZO4y5ggnzD9u:b-yraz1u) ([raw file](https://github.com/robotlocomotion/drake/blob/cd166cfb24dc56df72bd71a648fddfe5179545ff/bindings/pydrake/util/test/wrap_function_test.cc#L43)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, no need to have `EXPECT_EQ` for this one and the following one?
</blockquote></details>

OK There's no value returned by these lambdas, so no, not at the moment. I could use `check_signature` here, but I think that may not add too much value here.

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 147 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3jUGjQzqwoOQnP8aMz:-L3ne2rOHWnDVWxVzbxd:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d3e9d8a621d47135fd5cc3c4dc637fa176820cb3/bindings/pydrake/util/test/wrap_function_test.cc#L147)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Not sure how this comment is related to the following code.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/test/wrap_function_test.cc, line 251 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7841#-L3nC8j_GSUUVOnGS7Wn:-L3ndQuhKZDti5KVehcJ:b41lmwf) ([raw file](https://github.com/robotlocomotion/drake/blob/d3e9d8a621d47135fd5cc3c4dc637fa176820cb3/bindings/pydrake/util/test/wrap_function_test.cc#L251)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

:-) I've spent some time to figure out how it works (comma operator).
</blockquote></details>

OK Haha, yeah, I remember  being confused looking at this construct on `cppreference`; however, I've found that I like this more than recursively evaluating parameter packs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7841)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7841)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7841)*
<!-- Sent from Reviewable.io -->
